### PR TITLE
Transition plan for modularization

### PR DESCRIPTION
**Pull request by pbochynski**
**Description**

Changes proposed in this pull request:

- extend modularization concept with transition plan

**Related issue(s)**
- https://github.com/kyma-project/kyma/issues/13586
- https://github.com/kyma-project/kyma/issues/13176

### CODE REVIEW

File: concepts/modularization/README.md [hunk 1]
```diff
@@ -151,7 +152,7 @@ Yes, but under the hood, `kyma-operator` will be used to install component opera
 
 ## I have a simple component with a Helm chart. Why do I need an operator?
 
-With the operator, you can fully control your component lifecycle and ensure that your component are reconciled  to the desired state (watch component config and actual state).
+With the operator, you can fully control your component lifecycle and ensure that your component are reconciled to the desired state (watch component config and actual state). Each operator comes with a custom resource that describes module configuration and represents module installation status. It is a way to enable users with providing chart overrides in a controlled way.
```
Comment (jakobmoellersap) [inline concepts/modularization/README.md:155]:
```suggestion
With the operator, you can fully control your component lifecycle and ensure that your component are reconciled to the desired state (watch component config and actual state). Each operator comes with a custom resource that describes module configuration and represents module installation status. It is a way to enable users with providing chart overrides in a controlled way.
```
I think this is not only used for chart overrides but also for operator-related settings right? For sure, chart overrides are the most straight-forward use case, but also things like namespace injection for istio would be configured here right?

File: concepts/modularization/transition.md [hunk 2]
```diff
@@ -0,0 +1,30 @@
+# Transition from reconciler to kyma-operator
+
+With almost 20 modules and several teams involved it is not feasible to coordinate the transition to the new modular architecture in the single release. The plan is to switch modules to the new architecture gradually, avoiding big bang releases and keeping existing clusters stable. 
+
+## Phase 1 - local development
+
+You can start developing module operator with the provided guide, tools and libraries. Following the guide you will generate the operator scaffold, copy your helm chart inside and use the [`manifest` package](https://github.com/kyma-project/manifest-operator#use-manifest-packages-in-your-own-operator) to install it in your reconciliation logic. You can add configuration options to the spec of your operator CRD and use it as overrides for helm charts or in your custom logic (you still can use custom code in the operator).
```
Comment (jakobmoellersap) [inline concepts/modularization/transition.md:7]:
```suggestion
You can start developing a module operator with the provided guide, tools and libraries. Following the guide, you will generate the module and operator scaffold, copy your helm chart inside and use the [`manifest` package](https://github.com/kyma-project/manifest-operator#use-manifest-packages-in-your-own-operator) to install it in your reconciliation logic. You can add configuration options to the spec of your operator CRD and use it as overrides for helm charts or in your custom logic (you still can use custom code in the operator). This also allows you to differentiate different installation profiles and code paths.
```

File: concepts/modularization/transition.md [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+You can test operator on its own by applying operator CRD and creating the instance (you can install operator on your test cluster or even run it locally). When your operator works you can build the image and publish it (using the recommended pipeline template) and create a `ModuleTemplate` that contains a default custom resource used to install the module and operator deployment. You create one module template per release channel you want to support. Having `ModuleTemplate` you can start testing your module with `kyma-operator` . To do that you need to install kyma-operator and its custom resources in your test cluster, create a `Kyma` resource and add your module in the spec. `Kyma-operator` will install your operator CRD, will deploy your operator and start module installation by creating default custom resource. When your operator changes the status.state field it should be propagated to the `Kyma` resource status. You can install `kyma-operator` and enable/disable modules using `kyma CLI` or `kyma Dashboard`.
```
Comment (jakobmoellersap) [inline concepts/modularization/transition.md:8]:
```suggestion
You can test the operator on its own by applying the operator CRD and starting it independently (you can install operator on your test cluster or even run it locally). You can build the operator image and publish it (using the recommended pipeline template) and generate a `ModuleTemplate` that contains a default custom resource used to install the module and operator deployment. You create one `ModuleTemplate` per release channel you want to support. 

With the `ModuleTemplate`, you can integrate your module with `kyma-operator` . To do that, you need to install kyma-operator and its custom resources in your test cluster, create a `Kyma` resource and add your module in the spec. `Kyma-operator` will install your operator CRD, will deploy your operator and start the module installation by creating the default custom resource. When your operator changes the status.state field (for an example, take a look at the [template-operator](https://github.com/kyma-project/kyma-operator/blob/main/samples/template-operator/api/v1alpha1/mockup_types.go#L38-L54)) it should be propagated to the `Kyma` resource status. You can install `kyma-operator` and will be able to enable/disable modules using `kyma CLI` or `kyma Dashboard`.
```

**Review by jakobmoellersap — COMMENTED** [review]
Some smaller adjustments but overall LGTM. I think I would like to include more reference guides on Phase 1 for testing the integration with Kyma Operator. Will submit more info on that once we have that available

**Review by valentinvieriu — CHANGES_REQUESTED** [review]
I'm missing the mention that commponents that need a UI inside Busola they need a Busola Configmmap.
I'm not sure if this is an artifact or just needs to be added as one of the installation items, but needs to somehow be mentioned

File: concepts/modularization/transition.md [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+
+## Phase 2 - first module managed by `kyma-operator` integrated with KEB
+
+### Phase 2a - new clusters only (internal usage)
+
+`kyma-operator` is deployed in the control-plane and integrated with Kyma Environment Broker (KEB). Additional plan (preview) is created in KEB that creates Kyma clusters with kyma-operator only (no call to reconciler). This plan is the first playground to test KEB integration without affecting existing clusters and flows. The plan is tested with all environments (dev, stage, prod) to ensure we have all major integration issues resolved. The first module (frontrunner) can be submitted by the team and installed in the cluster. The minimal submission process is described and fully automated (manual steps are allowed only before submission). The minimal submission process looks like:
+- team verifies on their own if new module (or version) is product standard compliant
```
Comment (jakobmoellersap) [inline concepts/modularization/transition.md:15]: how do they do this?

File: concepts/modularization/transition.md [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+
+## Phase 2 - first module managed by `kyma-operator` integrated with KEB
+
+### Phase 2a - new clusters only (internal usage)
+
+`kyma-operator` is deployed in the control-plane and integrated with Kyma Environment Broker (KEB). Additional plan (preview) is created in KEB that creates Kyma clusters with kyma-operator only (no call to reconciler). This plan is the first playground to test KEB integration without affecting existing clusters and flows. The plan is tested with all environments (dev, stage, prod) to ensure we have all major integration issues resolved. The first module (frontrunner) can be submitted by the team and installed in the cluster. The minimal submission process is described and fully automated (manual steps are allowed only before submission). The minimal submission process looks like:
+- team verifies on their own if new module (or version) is product standard compliant 
+- team creates pull request to the control-plane that contains new/updated ModuleTemplate (by doing it team declares product standard compliance and functional correctness - can be checkbox in PR template)
```
Comment (jakobmoellersap) [inline concepts/modularization/transition.md:16]:
```suggestion
- team creates pull request to the control-plane that contains new/updated ModuleTemplate(s) (by doing it team declares product standard compliance and functional correctness - can be checkbox in PR template)
```
we have 3 ModuleTemplates for the different release channels

File: concepts/modularization/transition.md [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+- automated test verifies if the module operator can be installed/upgraded without issues for all release channels (with default settings) - it is not functional test or integration test 
+- automated test verifies if all images installed by operator are declared in the module template, signed by approved CA, vulnerability free
```
Comment (jakobmoellersap) [inline concepts/modularization/transition.md:18]:
```suggestion
- automated test verifies if all images installed by operator are declared in the module template, signed by approved CA, vulnerability free
```
I checked with @clebs and we want to check if it makes sense bundling the ModuleTemplate Information into the existing built module. That way we have only one image and all the checks can be done at build time so we could skip this

File: concepts/modularization/transition.md (hunk 4 above)
Comment (pbochynski) [reply concepts/modularization/transition.md:15]: We do not imply how to do that in the modularization concept. It is just a clear statement that it is the team's responsibility to deliver modules with proper quality. We can't validate it.

File: concepts/modularization/transition.md (hunk 6 above)
Comment (pbochynski) [reply concepts/modularization/transition.md:18]: We have to discuss it. Example:
Eventing operator installs eventing-controller and eventing-publishing-proxy. These 2 components are not directly installed by the kyma-operator, but should be signed and vulnerability free.

**Comment by pbochynski** [comment]
> I'm missing the mention that commponents that need a UI inside Busola they need a Busola Configmmap. I'm not sure if this is an artifact or just needs to be added as one of the installation items, but needs to somehow be mentioned

Done: https://github.com/kyma-project/community/pull/701/commits/f7aadf6b089e7a4ff2f77e2d61068091b116ad76

File: concepts/modularization/README.md (hunk 1 above)
Comment (NHingerl) [inline concepts/modularization/README.md:155]:
```suggestion
With the operator, you can fully control your component lifecycle and ensure that your component are reconciled to the desired state (watch component config and actual state). Each operator comes with a custom resource that describes the module configuration and represents the module installation status. It is a way to enable users with providing chart overrides in a controlled way.
```

File: concepts/modularization/README.md [hunk 7]
```diff
@@ -170,3 +171,7 @@ Use release channels to push the new version in the rapid channel first. After s
 ## Can I run multiple versions of central operator
 
 Yes. But to avoid concurrent updates and unpredictable outcomes, you must ensure that each module instance described by your module custom resource is reconciled (managed) by a single operator. You can achieve that by marking your custom resources with labels pointing to release channels in your module template. Then you can deploy one central module per release channel and update them independently.
+
+## How do we migrate all the modules to the new concept?
+
+The initial plan is described in the [transition](transition.md) document.
```
Comment (NHingerl) [inline concepts/modularization/README.md:177]:
```suggestion
Read about the initial plan in the [transition document](transition.md).
```

File: concepts/modularization/transition.md [hunk 8]
```diff
@@ -0,0 +1,32 @@
+# Transition from reconciler to kyma-operator
+
+With almost 20 modules and several teams involved it is not feasible to coordinate the transition to the new modular architecture in the single release. The plan is to switch modules to the new architecture gradually, avoiding big bang releases and keeping existing clusters stable.
```
Comment (NHingerl) [inline concepts/modularization/transition.md:3]:
```suggestion
With almost 20 modules and several teams involved, it is not feasible to coordinate the transition to the new modular architecture in a single release. The plan is to switch modules to the new architecture gradually, avoiding big bang releases and keeping existing clusters stable. 
```

**Review by valentinvieriu — APPROVED** [review]

File: concepts/modularization/transition.md (hunk 2 above)
Comment (NHingerl) [reply concepts/modularization/transition.md:7]: following up on Jakob's comment:

```suggestion
You can start developing a module operator with the provided guide, tools, and libraries. Following the guide, you generate the module and operator scaffold, copy your Helm chart inside, and use the [`manifest` package](https://github.com/kyma-project/manifest-operator#use-manifest-packages-in-your-own-operator) to install it in your reconciliation logic. You can add configuration options to the spec of your operator CRD and use them as overrides for Helm charts or in your custom logic (you still can use custom code in the operator). This way, you can also differentiate several installation profiles and code paths.
```

I don't love "allow" for anything that wasn't forbidden before. Usually it's  better to say "enables you to", or, simpler "you can".

File: concepts/modularization/transition.md [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+
+## Phase 1 - local development
+
+You can start developing module operator with the provided guide, tools and libraries. Following the guide you will generate the operator scaffold, copy your helm chart inside and use the [`manifest` package](https://github.com/kyma-project/manifest-operator#use-manifest-packages-in-your-own-operator) to install it in your reconciliation logic. You can add configuration options to the spec of your operator CRD and use it as overrides for helm charts or in your custom logic (you still can use custom code in the operator).
+You can test the operator on its own by applying the operator CRD and starting it independently (you can install operator on your test cluster or even run it locally). You can build the operator image and publish it (using the recommended pipeline template) and generate a `ModuleTemplate` that contains a default custom resource used to install the module and operator deployment. You create one `ModuleTemplate` per release channel you want to support.
```
Comment (NHingerl) [inline concepts/modularization/transition.md:8]:
```suggestion
To test the operator on its own, apply the operator CRD and start it independently. You can install the operator on your test cluster or even run it locally.  
You can build the operator image and publish it (using the recommended pipeline template), and generate a `ModuleTemplate` that contains a default custom resource used to install the module and operator deployment. For every release channel you want to support, create one `ModuleTemplate`. 
```

File: concepts/modularization/transition.md [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+
+With the `ModuleTemplate`, you can integrate your module with `kyma-operator` . To do that, you need to install kyma-operator and its custom resources in your test cluster, create a `Kyma` resource and add your module in the spec. `Kyma-operator` will install your operator CRD, will deploy your operator and start the module installation by creating the default custom resource. When your operator changes the status.state field (for an example, take a look at the [template-operator](https://github.com/kyma-project/kyma-operator/blob/main/samples/template-operator/api/v1alpha1/mockup_types.go#L38-L54)) it should be propagated to the `Kyma` resource status. You can install `kyma-operator` and will be able to enable/disable modules using `kyma CLI` or `kyma Dashboard`.
```
Comment (NHingerl) [inline concepts/modularization/transition.md:10]:
```suggestion
With the `ModuleTemplate`, you can integrate your module with `kyma-operator`: Install kyma-operator and its custom resources in your test cluster, create a `Kyma` resource, and add your module in the spec. `Kyma-operator` installs your operator CRD, deploys your operator, and starts the module installation by creating the default custom resource. When your operator changes the status.state field (for an example, take a look at the [template-operator](https://github.com/kyma-project/kyma-operator/blob/main/samples/template-operator/api/v1alpha1/mockup_types.go#L38-L54)), it should be propagated to the `Kyma` resource status. You can install `kyma-operator` and will be able to enable and disable modules using `Kyma CLI` or `Kyma Dashboard`.
```

What does "should" mean here? Is it a recommendation? Are there other options (then use "can")? Is it mandatory/necessary (then use "must))? What if it doesn't propagate the status information? Could we simply drop the "should" and write "..., it is propagated to the `Kyma` resource status."

File: concepts/modularization/transition.md [hunk 11]
```diff
@@ -0,0 +1,32 @@
...
+
+## Phase 2 - first module managed by `kyma-operator` integrated with KEB
+
+### Phase 2a - new clusters only (internal usage)
+
+`kyma-operator` is deployed in the control-plane and integrated with Kyma Environment Broker (KEB). Additional plan (preview) is created in KEB that creates Kyma clusters with kyma-operator only (no call to reconciler). This plan is the first playground to test KEB integration without affecting existing clusters and flows. The plan is tested with all environments (dev, stage, prod) to ensure we have all major integration issues resolved. The first module (frontrunner) can be submitted by the team and installed in the cluster. The minimal submission process is described and fully automated (manual steps are allowed only before submission). The minimal submission process looks like:
```
Comment (NHingerl) [inline concepts/modularization/transition.md:16]:
```suggestion
`kyma-operator` is deployed in the control-plane and integrated with Kyma Environment Broker (KEB). As a first playground to test KEB integration without affecting existing clusters and flows, an additional plan called "preview" is created in KEB. This plan creates Kyma clusters with kyma-operator only and does not call reconciler.
The plan is tested with all environments (dev, stage, prod) to ensure we have all major integration issues resolved. The first module (frontrunner) can be submitted by the team and installed in the cluster. The minimal submission process is described and fully automated (manual steps are allowed only before submission). The minimal submission process looks like:
```

File: concepts/modularization/transition.md [hunk 12]
```diff
@@ -0,0 +1,32 @@
...
+- team verifies on their own if new module (or version) is product standard compliant
```
Comment (NHingerl) [inline concepts/modularization/transition.md:17]: I guess the order of the steps matters, so we should use a numbered list.

```suggestion
1. The team verifies on their own if the new module (or version) complies with the SAP Product Standard.
```

Can we link to the relevant SAP Product Standard(s) here? Or do you refer to other product standards?

File: concepts/modularization/transition.md [hunk 13]
```diff
@@ -0,0 +1,32 @@
...
+
+## Phase 2 - first module managed by `kyma-operator` integrated with KEB
+
+### Phase 2a - new clusters only (internal usage)
+
+`kyma-operator` is deployed in the control-plane and integrated with Kyma Environment Broker (KEB). Additional plan (preview) is created in KEB that creates Kyma clusters with kyma-operator only (no call to reconciler). This plan is the first playground to test KEB integration without affecting existing clusters and flows. The plan is tested with all environments (dev, stage, prod) to ensure we have all major integration issues resolved. The first module (frontrunner) can be submitted by the team and installed in the cluster. The minimal submission process is described and fully automated (manual steps are allowed only before submission). The minimal submission process looks like:
+- team verifies on their own if new module (or version) is product standard compliant 
+- team creates pull request to the control-plane that contains new/updated ModuleTemplate(s) (by doing it team declares product standard compliance and functional correctness - can be checkbox in PR template)
```
Comment (NHingerl) [inline concepts/modularization/transition.md:18]:
```suggestion
2. The team creates a pull request to the control-plane that contains one or more new or updated ModuleTemplates. With this action, the team declares product standard compliance and functional correctness; for example, with a checkbox in the PR template.
```

File: concepts/modularization/transition.md [hunk 14]
```diff
@@ -0,0 +1,32 @@
...
+- automated test verifies if the module operator can be installed/upgraded without issues for all release channels (with default settings) - it is not functional test or integration test
```
Comment (NHingerl) [inline concepts/modularization/transition.md:19]:
```suggestion
3. An automated test verifies if the module operator can be installed or upgraded without issues for all release channels (with default settings). This is not a functional test or integration test!
```

File: concepts/modularization/transition.md (hunk 6 above)
Comment (NHingerl) [reply concepts/modularization/transition.md:20]:
```suggestion
4. An automated test verifies if all images installed by operator are declared in the module template, signed by approved CA, and free of vulnerabilities.
```

File: concepts/modularization/transition.md [hunk 15]
```diff
@@ -0,0 +1,32 @@
...
+- team verifies on their own if new module (or version) is product standard compliant 
+- team creates pull request to the control-plane that contains new/updated ModuleTemplate(s) (by doing it team declares product standard compliance and functional correctness - can be checkbox in PR template)
+- automated test verifies if the module operator can be installed/upgraded without issues for all release channels (with default settings) - it is not functional test or integration test 
+- automated test verifies if all images installed by operator are declared in the module template, signed by approved CA, vulnerability free
+- control-plane maintainer (codeowner) approves the PR which is automatically merged and rolled out
```
Comment (NHingerl) [inline concepts/modularization/transition.md:21]:
```suggestion
5. The control-plane maintainer (code owner) approves the PR, which is automatically merged and rolled out.
```

File: concepts/modularization/transition.md [hunk 16]
```diff
@@ -0,0 +1,32 @@
...
+
+
+### Phase 2b - move first module (frontrunner) from reconciler to kyma-operator (production)
+
+Regular plans use reconciler and kyma-operator in parallel (clusters are managed by both components, but their lists of modules don't overlap). Frontrunner module is disabled in the reconciler (ignored / replaced by dummy reconciler). `Kyma` resources containing the first module are populated for all existing clusters using script or KEB functionality. In each cluster `kyma-operator` installs frontrunner module operator with its CRD (configuration specification), and creates default configuration resource for the module. Module operator installs all needed components (including UI config map for Kyma Dashboard) and updated status of configuration custom resource (status.state = Ready). From that point in time cluster admin (or any user with sufficient roles) can edit the configuration resource to change module settings or can remove the module completely by deleting it from the list in the `Kyma` resource.
```
Comment (NHingerl) [inline concepts/modularization/transition.md:26]: Are "frontrunner" and "first" synonyms here? If yes, I suggest to keep just one term for better consistency.

```suggestion
Regular plans use reconciler and kyma-operator in parallel (clusters are managed by both components, but their lists of modules don't overlap). The frontrunner module is disabled in the reconciler (ignored or replaced by a dummy reconciler). `Kyma` resources containing the frontrunner module are populated for all existing clusters using a script or KEB functionality. In each cluster, `kyma-operator` installs the frontrunner module operator with its CRD (configuration specification), and creates a default configuration resource for the module. Module operator installs all needed components (including UI config map for Kyma Dashboard) and updates the status of the configuration custom resource (status.state = Ready). From that point in time, the cluster admin (or any user with sufficient authorization) can edit the configuration resource to change the module settings, or can remove the module completely by deleting it from the list in the `Kyma` resource.
```

I'm not sure I got the "updated" vs "updates" suggestion correctly; this sentence was a bit confusing.

File: concepts/modularization/transition.md [hunk 17]
```diff
@@ -0,0 +1,32 @@
...
+
+The submission process is extended with all aspects required for production deployment, like support, commercialization, security, micro-delivery, etc.
+## Phase 3 - all modules are managed by `kyma-operator` 
+
+All modules are enabled in `kyma-operator` and modules are added to Kyma resources for existing clusters. Reconciler is shutdown and KEB plans integrate only with `kyma-operator` (preview plan is also gone). All teams have submitted their modules using full version of submission process - the minimal submission process introduced in phase 2 extended to cover other aspects like support, microdelivery, etc. Other SAP teams can also submit their modules following improved technical guides and submission process description.
```
Comment (NHingerl) [inline concepts/modularization/transition.md:31]:
```suggestion
All modules are enabled in `kyma-operator`, and modules are added to Kyma resources for existing clusters. Reconciler is shut down, and KEB plans integrate only with `kyma-operator`. The "preview" plan is also gone. All teams have submitted their modules following the full submission process - the minimal submission process introduced in phase 2 extended to cover other aspects like support, microdelivery, etc. Other SAP teams can also submit their modules following improved technical guides and the full submission process. 
```

**Review by NHingerl — APPROVED** [review]

**Review by NHingerl — APPROVED** [review]

**Pull request merged**